### DISCUSSION TITLE
Heavily controlled form `onChange` handler - or, memoized values for `watch`

### DISCUSSION THREAD

**Discussion by kadoshms** [react-hook-form/react-hook-form#6278] [General]
Hey, first wanted to say thanks for this great library.

My form consists of only controlled components - multi selects, auto completers, and other friends.
In addition, changes to a specific section (field array) should trigger a server call, and then do some fun stuff in the view.

My common sense says that I should listen to changes from the form.
So I reached to `watch` and watched this session.
It looked something like that:

```
const section = watch('mySection');
useEffect(() => {
  // do some sever stuff, and then set state to re-render
}, [section]);
```

What I was surprised to see, is that the `section` value is changing on every render, causing an infinite loop and calls to the server.
I started digging and saw a place where it was suggested to use an `onChange` handler on the form.
The problem is that my components are complex, and controlled. so doing this on each on of them is totally a mess.

Is there a way to make the `watch` output change only when somehting really changed?
Or on the other hand, is there a way to have a general "onChange" event for all the fields under the form context?

Thanks!

**Comment by milkanas** [comment]
We had a similar issue, making an **onChange** handler can be very handy.

**Comment by bluebill1049** [comment]
`watch` doesn't memo the output. watch meant to be used for something in side the render, fast ana simple process.

**Reply by kadoshms** [reply]
I understand that, although I believe this is could be a very common case.
Is there any way I can achieve such behaviour?

**Reply by bluebill1049** [reply]
i believe if you only want to subscribe for changes, then 

- onChange
- 
```
watch(() => {
  // will return updated value
})
```

**Reply by eliseumds** [reply]
@bluebill1049 just a complement for the curious ones:

```ts
useEffect(() => {
  const subscription = watch(values => {
    // do something with values, like triggering an onChange callback
  });

  return () => {
    subscription.unsubscribe();
  };
}, [watch]);
```
